invokeai/backend/stable_diffusion/diffusers_pipeline.py

```diff
@@ -439,6 +445,74 @@ def generate_latents_from_embeddings(
         if timesteps.shape[0] == 0:
             return latents, attention_map_saver
 
+        # print("ip_adapter_image: ", type(ip_adapter_image))
+        if ip_adapter_data is not None and len(ip_adapter_data) > 0:
+            ip_adapter_info = ip_adapter_data[0]
+            ip_adapter_image = ip_adapter_info.image
+            # initialize IPAdapter
+            print("   width:", ip_adapter_image.width, " height:", ip_adapter_image.height)
+            # FIXME:
+            #   WARNING!
+            #   IPAdapter constructor modifies UNet model in-place
+            #   Adds additional cross-attention layers to UNet model for image embedding
+            #       need to figure out how to only do this if UNet hasn't already been modified by prior IPAdapter
+            #       and how to undo if ip_adapter_image is removed
+            #   Should reimplement to use existing model management context etc.
+            #
+            if "sdxl" in ip_adapter_info.ip_adapter_model:
```

Comment:
As noted in the FIXME, this needs to be handled properly through the model manager.

My 2-cents:

From a model management perspective, IPAdapters are similar to LoRAs in that they perform invasive modification of the UNet model. I think it makes sense to have IP-Adapter patching mirror our LoRA patching. In the initial version, we would incur the cost of patching/unpatching on every generation. But, this is a known issue that we are planning to solve for LoRAs in the near-term, so we could solve it for IP-Adapters at the same time.